fe/fe-core/src/main/java/org/apache/doris/planner/JoinCostEvaluation.java

```diff
@@ -147,7 +149,7 @@ public long constructHashTableSpace() {
                 Math.pow(1.5, (int) ((Math.log((double) rhsTreeCardinality/4096) / Math.log(1.5)) + 1)) * 4096;
         double nodeOverheadSpace = nodeArrayLen * 16;
         double nodeTuplePointerSpace = nodeArrayLen * rhsTreeTupleIdNum * 8;
-        return Math.round((bucketPointerSpace + (double) rhsTreeCardinality * rhsTreeAvgRowSize
+        return Math.round((bucketPointerSpace + (double) rhsTreeCardinality * rhsTreeAvgRowSize * COMPRESSION_RATIO
```

Comment:
yes, you are right. multiplying by a fixed ratio is not accurate. But the error in data size is not only just introduced by compressed ratio.
Since, we don't have accurate statistics now, the average row size and cardinality are also inaccurate.
So the compression ratio of 5 is used here, based on observations of imported data to Doris.

Reply:
In fact, when we wrote this cost formula, we tested it with real memory consumption. According to the previous test, the current formula is relatively accurate.

If you have actually tested the gap between cost and the real memory consumption, can you provide the test data. Let's analyze where the specific gap is, instead of simply multiplying by a fixed value.

Reply:
The `avgRowSize` statistics caused by the compression ratio are inaccurate,
I understand that corrections should be made in `computeStats` of OlapScanNode, not in JoinCostEvaluation, until more accurate statistics collection is achieved.

![image](https://user-images.githubusercontent.com/13197424/160399501-24112608-51c8-45a1-9d64-417285ef029d.png)

Reply:
> In fact, when we wrote this cost formula, we tested it with real memory consumption. According to the previous test, the current formula is relatively accurate.
> 
> If you have actually tested the gap between cost and the real memory consumption, can you provide the test data. Let's analyze where the specific gap is, instead of simply multiplying by a fixed value.

we do test on tpc-h with sf = 100 and 3 be with 64GB memory.
q9, q12 and q21 failed on previous cost formula.

Reply:
> The `avgRowSize` statistics caused by the compression ratio are inaccurate, I understand that corrections should be made in `computeStats` of OlapScanNode, not in JoinCostEvaluation, until more accurate statistics collection is achieved.
> 
> `totalBytes` is the compressed size. ![image](https://user-images.githubusercontent.com/13197424/160399501-24112608-51c8-45a1-9d64-417285ef029d.png)

i agree with that

Reply:
I checked the current average row size of the code into the following cases:
1. OlapScanNode: avgRowSize=totalBytes (compressed) / cardinality
2. Agg or Sort: avgRowSize= the actual sum of the size of each column
(not sure if there are other possible values)

So
1. Is it possible to use the actual row size directly into the current formula to match the actual consumption? If yes, then I'm in favor of changing the row size. If not, are there other problems with the current formula?
3. Some operators do not actually use the compressed size, so it is best to modify the average size of the row in PlanNode.computeStats()

Reply:
> q9, q12 and q21 failed on previous cost formula.

The above 3 queries cause BE crash because of RPC exception, not hash table size. So do we have to consider the RPC  timeout? In fact, the memory is enough in our test environment.

Reply:
> I checked the current average row size of the code into the following cases:
> 
> 1. OlapScanNode: avgRowSize=totalBytes (compressed) / cardinality
> 2. Agg or Sort: avgRowSize= the actual sum of the size of each column
>    (not sure if there are other possible values)
> 
> So
> 
> 1. Is it possible to use the actual row size directly into the current formula to match the actual consumption? If yes, then I'm in favor of changing the row size. If not, are there other problems with the current formula?
> 2. Some operators do not actually use the compressed size, so it is best to modify the average size of the row in PlanNode.computeStats()

I agree with u. This is same suggestion with @[xinyiZzz](https://github.com/xinyiZzz).

Reply:
> > q9, q12 and q21 failed on previous cost formula.
> 
> The above 3 queries cause BE crash because of RPC exception, not hash table size. So do we have to consider the RPC timeout? In fact, the memory is enough in our test environment.

indeed, sometimes failed with RPC timeout, sometimes failed with OOM

Reply:
One more thing I need to remind. The current OOM has a situation as following:

There are multiple hash join nodes in one Plan Fragment. The memory of each single hash join node is not exceeded. But the memory of the total plan fragment is overrun.
I'm not sure if this is the case in your test.
If there is, then it's not really a matter of join memory evaluation. Rather, we cannot simply evaluate the memory problems of plan fragments.

Reply:
> One more thing I need to remind. The current OOM has a situation as following:
> 
> There are multiple hash join nodes in one Plan Fragment. The memory of each single hash join node is not exceeded. But the memory of the total plan fragment is overrun. I'm not sure if this is the case in your test. If there is, then it's not really a matter of join memory evaluation. Rather, we cannot simply evaluate the memory problems of plan fragments.

yes, take account continuous join memory together is next step i want to do

Reply:
> yes, take account continuous join memory together is next step i want to do

If you plan to do this later, we can discuss it~
I am also thinking about how to estimate the memory consumption of Fragment and take admission control during query planning.

Reply:
> 

i'm glad to discuss it, and if you already have some plans in this domain, i could do sometimes else